bin/servicerouter.py

```diff
@@ -823,64 +824,60 @@ def compareWriteAndReloadConfig(config, config_file):
 
 
 def get_health_check(app, portIndex):
-    for check in app.app['healthChecks']:
+    for check in app['healthChecks']:
         if check['portIndex'] == portIndex:
             return check
     return None
 
 
 def get_apps(marathon):
-    tasks = marathon.tasks()
+    apps = marathon.list()
+    logger.debug("got apps %s", map(lambda app: app["id"], apps))
 
-    apps = dict()
+    marathon_apps = dict()
 
-    for task in tasks:
-        # For each task, extract the app it belongs to and add a
-        # backend for each service it provides
-        if 'servicePorts' not in task:
-            continue
+    for app in apps:
+        appId = app['id']
 
-        for i in xrange(len(task['servicePorts'])):
+        marathon_app = MarathonApp(marathon, appId, app)
+        if 'HAPROXY_GROUP' in marathon_app.app['env']:
+                    marathon_app.groups = \
+                            marathon_app.app['env']['HAPROXY_GROUP'].split(',')
+        marathon_apps[appId] = marathon_app
+
+        for task in app['tasks']:
             # Marathon 0.7.6 bug workaround
             if len(task['host']) == 0:
                 logger.warning("Ignoring Marathon task without host " +
                                task['id'])
                 continue
 
-            servicePort = task['servicePorts'][i]
-            port = task['ports'][i] if len(task['ports']) else servicePort
-            appId = task['appId']
-
-            if appId not in apps:
-                app_tmp = MarathonApp(marathon, appId)
-                if 'HAPROXY_GROUP' in app_tmp.app['env']:
-                    app_tmp.groups = \
-                            app_tmp.app['env']['HAPROXY_GROUP'].split(',')
-                apps[appId] = app_tmp
-
-            app = apps[appId]
-            if servicePort not in app.services:
-                app.services[servicePort] = MarathonService(
-                    appId, servicePort, get_health_check(app, i))
+            service_ports = app['ports']
+            for i in xrange(len(service_ports)):
+                servicePort = service_ports[i]
+                port = task['ports'][i] if len(task['ports']) else servicePort
+                if servicePort not in marathon_app.services:
```

Comment:
So how this was all setup because it did a task list then lazily fetched the apps is structurally a bit backwards from what it should look like now that we are starting at the "apps" and then iterating over the tasks inside (Wish I'd been aware apps had the tasks available if you asked right...).

I'm fine in this PR just merging with the old code structure since the speed improvement is so drastic and things work.

Longer term what should happen is there is a loop over the "ports" array in the base app, that builds out the set of services. Then in looping over the tasks just attach them to services, rather than doing the "we aren't sure if this port/service exists yet" check.

I think there may be some latent issues around deployments which will need to be sorted out (There are two versions of an app during a deployment with different sets of ports, ServiceRouter config, etc.). Fortunately most apps don't have changing sets of ports so it hasn't come up yet. (Switching to be mesos task label based rather than marathon could help there).